crates/bevy_render/src/picking.rs

```diff
@@ -0,0 +1,437 @@
+        texture: &Texture,
+        current_buffer_index: &CurrentBufferIndex,
+    ) {
+        let Some(buffers) = self.buffers.as_ref() else { return; };
+
+        // Copy current frame to next buffer
+        let copy_index = (current_buffer_index.0 + 1) % BUFFER_COUNT;
+        buffers.copy_texture_to_buffer(encoder, texture, copy_index);
+
+        // Map current buffer that will be copied and sent after the graph has finished
+        let map_index = current_buffer_index.0;
+        let buffer_slice = buffers.entity_buffers[map_index].slice(..);
+        let (tx, _) = self.map_status_channel.clone();
+        buffer_slice.map_async(MapMode::Read, move |result| match result {
+            Ok(_) => tx.try_send(()).unwrap(),
+            Err(err) => panic!("Failed to map entity buffer {map_index}: {err}"),
+        });
+    }
+}
+
+/// Data sent between the render world and main world
+#[derive(Default)]
+struct GpuPickingData {
+    /// Padding required to compute the entity with the exact position in the buffer
+    padded_bytes_per_row: usize,
+    /// Buffer representing the entity texture
+    mesh_id_buffer: Vec<u8>,
+    /// A list of the visible entities during the frame the buffer was generated
+    /// The buffer contains an index into this list
+    visible_mesh_entities: Vec<Entity>,
+}
+
+/// Contains the buffers and their dimensions required for gpu picking
+#[derive(Clone)]
+struct GpuPickingCameraBuffers {
+    entity_buffers: [Buffer; BUFFER_COUNT],
+    // All buffers have the same dimension so we only need one
+    buffer_dimensions: BufferDimensions,
+}
+
+impl GpuPickingCameraBuffers {
+    /// Copies the given texture to the buffer at the given index
+    fn copy_texture_to_buffer(
+        &self,
+        encoder: &mut CommandEncoder,
+        texture: &Texture,
+        buffer_index: usize,
+    ) {
+        encoder.copy_texture_to_buffer(
+            texture.as_image_copy(),
+            wgpu::ImageCopyBuffer {
+                buffer: &self.entity_buffers[buffer_index],
+                layout: ImageDataLayout {
+                    offset: 0,
+                    bytes_per_row: Some(self.buffer_dimensions.padded_bytes_per_row as u32),
+                    rows_per_image: None,
+                },
+            },
+            Extent3d {
+                width: self.buffer_dimensions.width as u32,
+                height: self.buffer_dimensions.height as u32,
+                ..default()
+            },
+        );
+    }
+}
+
+/// This is created every frame and contains a list of the currently visible entities
+#[derive(Component)]
+pub struct VisibleMeshEntities(pub Vec<Entity>);
+
+fn send_buffer(
+    query: Query<&ExtractedGpuPickingCamera>,
+    render_device: Res<RenderDevice>,
+    vis: Query<&VisibleMeshEntities>,
+    current_buffer_index: Res<CurrentBufferIndex>,
+) {
+    let visible_mesh_entities = vis
+        .get_single()
+        .expect("There should only be 1 VisibleMeshEntities")
```

Comment:
Should this be a `Resource`?

Reply:
Resource don't get cleared every frame and I'm just not sure if it may cause issues

Reply:
I switched it to a Res anyway, I feel like it's probably fine and it will be easy to fix if someone encounters an issue.